### ISSUE TITLE
RPDE: No deleted items?

### ISSUE THREAD

**Issue by GetTheData** [GLL-Better/opendata#13]
According the the [Realtime Paged Data Exchange](https://www.openactive.io/realtime-paged-data-exchange/), items which are deleted should have state=deleted

I have been polling your feed for a while, and have never seen an item where state=deleted.

Have you implemented this?

If no, what happens if an item is deleted? Is it just removed from the feed?

**Comment by itsterry** [comment]
I don't think we've deleted any since we started with the feed



----------------------------------------------
Terry Shuttleworth
Morse Digital LLP

terry@morsedigital.com

Voice
-----
Mobile: +44 (0) 77 68 91 81 93
Office: +44 (0) 20 7193 0909

IM
--
Google Chat: itsterry@gmail.com
Skype: itsterry

Company
-------
Registered in England No. OC368394
Vat No. GB 122 2042 79

**Comment by GetTheData** [comment]
@itsterry I seem to be seeing items which once existed, but no longer exist when I do a full crawl.

Example ID: 10297669

Is it possible that items are being removed from the feed?

**Comment by itsterry** [comment]
The example given is for Jan 3, 2018.

Items which are in the past by enough of a margin should no longer be in
the feed.

They're not necessarily deleted, but they're also no longe relevant, so we
don't include them



----------------------------------------------
Terry Shuttleworth
+44 (0) 77 68 91 81 93
Skype: itsterry
MSN: terry@shuttleworths.net

**Comment by GetTheData** [comment]
@itsterry thanks for your help with this!

Here is an example ID with a date in the future:

9766911

I'm seeing 1000s of these with dates in the future.

From the (RPDE)[https://www.openactive.io/realtime-paged-data-exchange/] spec:

> If any record is added to the list or updated it must remain in the list in perpetuity while it is in an "updated" state, or remain in the list for at least 7 days from the point in time at which it transitioned to a "deleted" state. This ensures that deletions are synchronised to data consumers (System 2).

I interpret that as an event needing to be in the deleted state for seven days before it can be dropped from the feed (even if startDate is in the past).

**Comment by itsterry** [comment]
Yes, that makes sense. I know how to implement this, and I'll get it done.

NB: we're in deployment lockdown till w/c April 16th, but I'll try to get it into the first release after that date

**Comment by GetTheData** [comment]
Thanks @itsterry - much appreciated!

Perhaps you could update this thread once released and I'll check my implementation still works ok.

**Comment by domfennell** [comment]
@GetTheData a little late to the party here! @itsterry, is there now an implementation to mark deleted items as `state=deleted`?

**Comment by itsterry** [comment]
Yes we did that, but it's still in a version awaiting approval. 

I'll see if I can move it up the schedule

**Comment by domfennell** [comment]
Excellent news, @itsterry! A nudge up the schedule would be very much appreciated! (cc @GetTheData)

Any idea roughly when this might be approved? Are we talking days, weeks, months...

**Comment by itsterry** [comment]
It's not publicly visible on better.org.uk, so our marketing team won't be able to approve it. It's only sitting in a branch awaiting approval as it was bundled in with a bunch of publicly visible changes. I'll see if I can get around that

**Comment by itsterry** [comment]
Right. I managed to extract that piece of code and hot-deploy it for you. Could you check it for me?

NB: there are a LOT of deleted records. You may want to skip straight to this page:
https://www.better.org.uk/odi/sessions.json?afterId=36190000

**Comment by domfennell** [comment]
Superb work, @itsterry! Really appreciate that!

I'll wait for the data to propagate through to our platform and let you know. Given the large number of records (as you intimated), it might take a day or so, after which I'll let you know. Of course, @GetTheData might be able to check it for you before then.

Thanks again for your efforts.

**Comment by itsterry** [comment]
No problem @domfennell . Will you close this if all ok?

**Comment by GetTheData** [comment]
@itsterry @domfennell I'll pull down the lot tomorrow and let you know what I find.

**Comment by GetTheData** [comment]
@itsterry you're right - lots of deleted items! Taking far longer than usual to do a full fetch, I'm going to leave it overnight and will update tomorrow. cc @domfennell

**Comment by GetTheData** [comment]
@itsterry @domfennell I did a full fetch overnight, and built a Data Profile from the results:

https://www.getthedata.com/openactive/gll/data-profile

You can see there are now 1,788,940 items, of which 1,070,661 are deleted.

718,279 items are updated, whereas a couple of days ago this was 784,902, so this doesn't seem unreasonable.

One issue remaining is that a deleted item does not take any data. From the spec:

> Deleted items are included in the response with a "deleted" state, but no &lt;data&gt; associated.

https://www.w3.org/2017/08/realtime-paged-data-exchange/#deleted-items

Here is an example from the spec:

```
{
  "state": "deleted",
  "kind": "session",
  "id": "{d97f73fb-4718-48ee-a6a9-9c7d717ebd85}",
  "modified": 1453931925
}
```

Here is what I currently see in the GLL feed:

```
{
            "state": "deleted",
            "kind": "session",
            "id": 17727567,
            "modified": "2018-01-11T12:49:07",
            "data": {
                "type": "Event",
                "url": "https:\/\/better.org.uk\/leisure-centre\/belfast\/brook-activity-centre\/timetable",
                "name": "Total Body Conditioning",
                "genderRestriction": "female",
                "location": {
                    "type": "SportsActivityLocation",
                    "name": "MAIN HALL",
                    "containedInPlace": {
                        "type": "Place",
                        "url": "https:\/\/better.org.uk\/leisure-centre\/belfast\/brook-activity-centre",
                        "name": "Brook Activity Centre",
                        "address": {
                            "type": "PostalAddress",
                            "addressLocality": "Dunmurry",
                            "addressRegion": "County Antrim",
                            "postalCode": "BT17 0RL",
                            "streetAddress": "25 Summerhill Road"
                        },
                        "geo": {
                            "latitude": 54.552,
                            "longitude": -6.0219
                        },
                        "telephone": "028 9560 9988"
                    }
                },
                "organizer": {
                    "type": "Organization",
                    "name": "Better"
                },
                "activity": {
                    "type": "Concept",
                    "notation": 16,
                    "prefLabel": "Fitness Classes",
                    "inScheme": "https:\/\/better.org.uk\/odi\/activities"
                },
                "eventStatus": "EventScheduled",
                "level": "All",
                "ageRange": "16-55",
                "contributor": {
                    "type": "Person",
                    "name": "FIONNUALA"
                },
                "description": "<p>Once you start, you won\u2019t ever want to stop. These intense, exciting classes combine cardio and conditioning for a full on, high-energy workout.<\/p>\r\n<p>Better for: Strength and conditioning.<\/p>\r\n<p>Approx. 420 calories per hour.<\/p>",
                "image": "https:\/\/assets.better.org.uk\/uploads\/asset\/attachment\/41037\/418963_GLL_AdHoc_Image_resize_750x330_Mar18.jpg",
                "offer": {
                    "url": "https:\/\/better.legendonlineservices.co.uk\/enterprise\/account\/Login",
                    "type": "Offer"
                },
                "startDate": "2018-11-06T09:45:00Z",
                "endDate": "2018-11-06T10:45:00Z",
                "duration": "PT60M",
                "@context": "https:\/\/www.openactive.io\/ns\/oa.jsonld"
            }
}
```
Note that a deleted item only needs to be present in the feed for seven days, after which it can be deleted:

> If any item is added to the list or updated it must remain in the list in perpetuity while it is in an "updated" state, or remain in the list for at least 7 days from the point in time at which it transitioned to a "deleted" state.

https://www.w3.org/2017/08/realtime-paged-data-exchange/#deleted-items

**Comment by itsterry** [comment]
That's done and live for you

**Comment by GetTheData** [comment]
Thanks @itsterry I can see the change in the Feed Explorer I built:

https://www.getthedata.com/openactive/gll/feed

I'll do another full fetch now and see what comes back.

cc: @domfennell

**Comment by lukehesluke** [comment]
Hi @itsterry - I work at imin with @domfennell . We've been re-syncing your feed and noticed that many of the Events have a startDate in the distant past.

Example: https://www.better.org.uk/odi/sessions.json?afterTimestamp=2018-11-24T13:13:19&afterId=35874649 includes deleted Events with `startDate` set to some time in the 1670s. I haven't done a more thorough analysis of this yet so I don't know if this affects "updated" Events. If it only affects "deleted" events then it's not as urgent an issue as it is if also affects "updated"s - still a bug nonetheless

**Comment by itsterry** [comment]
Yep: sometimes venues forget to put in the date for a series of sessions, or put it in a strange way, and the system uses a really old date. These never show as they're in the past, but they get soft-deleted by the nightly housekeeping task and should get hard-deleted 7 days later

**Comment by domfennell** [comment]
Thanks for your continued efforts, @itsterry. We'll take a look at the updated data in due course and let you know. We appreciate you keeping this issue open until it's completely resolved. Hopefully there now!

And look forward to hearing what you think, @GetTheData!

**Comment by GetTheData** [comment]
@itsterry @domfennell I've done a full refresh and all looks good to me.

I have 704,166 updated items, and 704,166 items with data, so that looks to be spot on.

Final question from me to @itsterry - are you expecting to remove the deleted items from the feed after a period of time?

According to the spec deleted items should be there for at least seven days. But I guess they could theoretically remain in place indefinitely, although this would be a bit unwieldy for the data consumer:

> If any item is added to the list or updated it must remain in the list in perpetuity while it is in an "updated" state, or remain in the list for at least 7 days from the point in time at which it transitioned to a "deleted" state.

https://www.w3.org/2017/08/realtime-paged-data-exchange/#deleted-items

**Comment by itsterry** [comment]
They *should* delete after 7 days. Do let me know if that isn't happening

**Comment by GetTheData** [comment]
Thank @itsterry - so you would expect the 842,363 deleted items I am currently seeing to have dramatically decreased by say next Friday? If so I'll make a diary entry to check. Cc: @domfennell

**Comment by domfennell** [comment]
Excellent. I'm away, but in my absence, a member of the imin team should be able to take a quick look at the end of next week. Cc @nishaldesai.

**Comment by nishaldesai** [comment]
@GetTheData how's this now looking your side re: the deleted items decreasing in number?

**Comment by GetTheData** [comment]
Hi @nishaldesai the deleted items are indeed decreasing significantly.

You can see this from our GLL data profile:

https://www.getthedata.com/openactive/gll/data-profile

Last week GLL had 1,070,661 deleted items.

Currently they have 249,067.

Cc: @itsterry @domfennell
